refactor: Paging for the search results and twig templates

aragon999 opened the pull request:
### 1. Why is this change necessary?
Just simplify the current setup to work with pagination.

### 2. What does this change do, exactly?
See changelog.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.


<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2390"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

aragon999 committed:
NEXT-21015 - Refactor: Paging for the search results and twig templates

shyim commented on changelog/_unreleased/2022-03-14-refactor-pagination-templates-and-paging-of-the-search-result.md:
```diff
@@ -0,0 +1,42 @@
+* Refactor `@Storefront/storefront/page/product-detail/review/review.html.twig` and adapt to changes for the pagination for the next major version v6.5.0
+* Refactor `@Storefront/storefront/page/account/order-history/index.html.twig` and use the already computed value of the `totalPages` of the orders search result and adapt to changes for the pagination for the next major version v6.5.0
+___
+# 6.5.0.0
```
```suggestion
# Next major version changes
```

shyim commented on src/Storefront/Framework/Page/StorefrontSearchResult.php:
```diff
@@ -2,9 +2,11 @@
 use Shopware\Core\Framework\DataAbstractionLayer\Search\EntitySearchResult;
 
+/**
+ * @deprecated tag:v6.5.0 - will be removed without replacement use `EntitySearchResult` instead, all methods are now contained in the `EntitySearchResult` and the sorting was not in use any more
```
```suggestion
 * @deprecated tag:v6.6.0 - will be removed without replacement use `EntitySearchResult` instead, all methods are now contained in the `EntitySearchResult` and the sorting was not in use any more
```

shyim commented on src/Core/Framework/DataAbstractionLayer/Search/EntitySearchResult.php:
```diff
@@ -168,6 +168,18 @@ public function setEntity(string $entity): void
+
+        //next page mode fetches only the next pages with, not the exact count
+        if ($this->getCriteria()->getTotalCountMode() === Criteria::TOTAL_COUNT_MODE_NEXT_PAGES) {
+            $total += $this->getCriteria()->getOffset();
```
Is not covered by tests

aragon999 replied:
It should be tested by: https://github.com/shopware/platform/pull/2390/files#diff-61eff07120696cdc628dee5f038aebf1fb58ecd30f80f3b2c0a925100648ad33R106-R113

But I moved the test file to the unit test directory (as it only contained unit tests).

shyim commented on src/Storefront/Resources/views/storefront/component/pagination.html.twig:
```diff
@@ -1,7 +1,12 @@
-    {% set totalPages = (totalEntities / criteria.limit)|round(0, 'ceil') %}
+    {% set currentPage = currentPage ?? entities.page %}
+    {# @deprecated tag:v6.5.0 - variable `totalEntities` will be removed and the variable `total` will be removed and not considered anymore, set `totalPages` instead #}
+    {% if feature('V6_5_0_0') %}
```
```suggestion
    {% if feature('V6_6_0_0') %}
```

shyim commented on src/Storefront/Resources/views/storefront/component/product/listing.html.twig:
```diff
@@ -30,10 +30,14 @@
-                                        entities: searchResult,
-                                        criteria: searchResult.criteria
-                                    } %}
+                                    {% if feature('V6_5_0_0') %}
```
```suggestion
                                    {% if feature('v6_6_0_0') %}
```

shyim replied:
Add there also an annotation that this if needs to be removed

shyim commented on src/Storefront/Resources/views/storefront/component/review/review.html.twig:
```diff
@@ -248,11 +248,15 @@
-                                                                criteria: criteria,
-                                                                total: productReviewCount,
-                                                            }  %}
+                                                            {% if feature('V6_5_0_0') %}
```
```suggestion
                                                            {% if feature('v6_6_0_0') %}
```

shyim replied:
Annotation

shyim commented on src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig:
```diff
@@ -60,10 +59,14 @@
-                                                entities: page.orders,
-                                                criteria: criteria
-                                            }  %}
+                                            {% if feature('V6_5_0_0') %}
```
```suggestion
                                            {% if feature('v6_6_0_0') %}
```

shyim replied:
Annotation

shyim commented on src/Storefront/Resources/views/storefront/page/product-detail/review/review.html.twig:
```diff
@@ -245,11 +245,16 @@
-                                                                total: productReviewCount,
-                                                            }  %}
+                                                            {# @deprecated tag:v6.5.0 - variable `productReviewCount` will not be considered anymore, change `totalPages` if needed #}
+                                                            {% if feature('V6_5_0_0') %}
```
```suggestion
                                                            {% if feature('v6_6_0_0') %}
```

shyim commented:
Sorry for the late response. I copied from review comments from our internal Gitlab

aragon999 commented:
I implemented the requested changes and added a comment on one thing which is not clear to me.

Sad that we now have to carry around the `StorefrontSearchResult` one major version longer :-(

dfrancos-hub commented:
hi @aragon999, please excuse us for the late response. If you plan to continue on this PR, could you resolve the conflicts? thanks in advance 🙏

aragon999 commented:
@dfrancos-hub yes, I mean I fixed and implemented all requested changes.

This PR has some deprecations for at the moment 6.6. As this PR has not been merged in the 6.5 cycle (why?), should these deprecations now target the 6.7 release? (Note that originally this PR was done for 6.4).

mitelg commented:
@aragon999 as we are already preparing the 6.6 release, your deprecations need to target 6.7.0.

sorry for the inconvenience

aragon999 committed:
NEXT-21015 - Fixup: Implement StorefrontSearchResult deprecations

aragon999 commented:
@dfrancos-hub @mitelg I have rebased the PR and adapted it to the changes in the meantime.

Furthermore I am not sure if the implementation of the depreciations, is good in this way, or not. Therefore I left them as additional commit.

mitelg commented on src/Storefront/Framework/Page/StorefrontSearchResult.php:
```diff
@@ -30,6 +33,11 @@ class StorefrontSearchResult extends EntitySearchResult
     public function getSortings(): array
     {
+        Feature::triggerDeprecationOrThrow(
+            'v6.6.0.0',
```
```suggestion
            'v6.7.0.0',
```

mitelg commented on src/Storefront/Framework/Page/StorefrontSearchResult.php:
```diff
@@ -38,25 +46,31 @@ public function getSortings(): array
     public function setSortings(array $sortings): void
     {
+        Feature::triggerDeprecationOrThrow(
+            'v6.6.0.0',
```
```suggestion
            'v6.7.0.0',
```

mitelg commented on src/Storefront/Framework/Page/StorefrontSearchResult.php:
```diff
@@ -38,25 +46,31 @@ public function getSortings(): array
     public function getSorting(): ?string
     {
+        Feature::triggerDeprecationOrThrow(
+            'v6.6.0.0',
```
```suggestion
            'v6.7.0.0',
```

mitelg commented on src/Storefront/Framework/Page/StorefrontSearchResult.php:
```diff
@@ -38,25 +46,31 @@ public function getSortings(): array
-            return 1;
-        }
+        Feature::triggerDeprecationOrThrow(
+            'v6.6.0.0',
```
```suggestion
            'v6.7.0.0',
```

mitelg commented on src/Storefront/Resources/views/storefront/component/pagination.html.twig:
```diff
@@ -1,7 +1,12 @@
-    {% set totalPages = (totalEntities / criteria.limit)|round(0, 'ceil') %}
+    {% set currentPage = currentPage ?? entities.page %}
+    {# @deprecated tag:v6.7.0 - variable `totalEntities` will be removed and the variable `total` will be removed and not considered anymore, set `totalPages` instead #}
+    {% if feature('v6_6_0_0') %}
```
```suggestion
    {% if feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/component/product/listing.html.twig:
```diff
@@ -1,5 +1,8 @@
-{% set currentPage = ((searchResult.criteria.offset + 1) / searchResult.criteria.limit )|round(0, 'ceil') %}
-{% set paginationConfig = { page: currentPage }|json_encode %}
+{# @deprecated tag:v6.7.0 - variable `currentPage` will be removed #}
+{% if not feature('v6_7_0_0') %}
```
```suggestion
{% if not feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/component/product/listing.html.twig:
```diff
@@ -32,10 +35,15 @@
-                                        criteria: searchResult.criteria
-                                    } %}
+                                    {# @deprecated tag:v6.7.0 - variable `criteria` will be removed #}
+                                    {% if feature('v6_7_0_0') %}
```
```suggestion
                                    {% if feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/component/product/listing.html.twig:
```diff
@@ -83,11 +91,19 @@
-                            paginationLocation: 'bottom',
-                        } %}
+                        {# @deprecated tag:v6.7.0 - variable `criteria` will be removed #}
+                        {% if feature('v6_7_0_0') %}
```
```suggestion
                        {% if feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/component/review/review.html.twig:
```diff
@@ -202,7 +202,10 @@
                                         {% block component_review_list_paging %}
-                                            {% set criteria = reviews.criteria %}
+                                            {# @deprecated tag:v6.7.0 - unused variable `criteria` will be removed, use reviews.criteria instead #} 
+                                            {% if not feature('v6_7_0_0') %}
```
```suggestion
                                            {% if not feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/component/review/review.html.twig:
```diff
@@ -239,11 +242,16 @@
-                                                                total: productReviewCount,
-                                                            } %}
+                                                            {# @deprecated tag:v6.7.0 - variable `productReviewCount` will not be considered anymore, change `totalPages` if needed #}
+                                                            {% if feature('v6_7_0_0') %}
```
```suggestion
                                                            {% if feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig:
```diff
@@ -40,9 +40,10 @@
-                            {% set criteria = page.orders.criteria %}
-
-                            {% set totalPages = (page.orders.total/criteria.limit)|round(0,'ceil') %}
+                            {% if not feature('v6_7_0_0') %}
```
```suggestion
                            {% if not feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig:
```diff
@@ -58,10 +59,16 @@
-                                            } %}
+
+                                            {# @deprecated tag:v6.7.0 - variable `criteria` will be removed #}
+                                            {% if feature('v6_7_0_0') %}
```
```suggestion
                                            {% if feature('v6.7.0.0') %}
```

mitelg commented on src/Storefront/Resources/views/storefront/page/product-detail/review/review.html.twig:
```diff
@@ -234,11 +234,16 @@
-                                                                total: productReviewCount,
-                                                            } %}
+                                                            {# @deprecated tag:v6.7.0 - variable `productReviewCount` will not be considered anymore, change `totalPages` if needed #}
+                                                            {% if feature('v6_7_0_0') %}
```
```suggestion
                                                            {% if feature('v6.7.0.0') %}
```

dfrancos-hub commented:
Hi @aragon999, could you adjust the changelog description to fix the PHPLint check?https://github.com/shopware/shopware/actions/runs/7170576701/job/19523620544?pr=2390 .
Also, the jest admin test is failing probably due to outdated branch, as it doesnt seem to be related to your changes. 
Thanks in advance 🙏

aragon999 committed:
NEXT-21015 - Fixup: Unify and use feature flag v6.7.0.0

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

aragon999 commented:
@dfrancos-hub I fixed the changelog lint.

But the admin jest tests are also failing on trunk, so I think this is unrelated.

ghost commented:
Hello,

thank you for creating this pull request.
I have opened an issue on our Issue Tracker for you. See the issue link: https://issues.shopware.com/issues/NEXT-32337

Please use this issue to track the state of your pull request.

dfrancos-hub commented on src/Storefront/Page/Product/Review/ReviewLoaderResult.php:
```diff
@@ -5,97 +5,197 @@
-#[Package('storefront')]
-class ReviewLoaderResult extends StorefrontSearchResult
-{
+if (Feature::isActive('v6.7.0.0')) {
```
Hi @aragon999, am pasting here comments from the internal core team review:

You can do this, but you shouldn't. The use of the FeatureFlags in this MR are not ideal, as they try to change everything completely on the point of v6.7, this will most likely fail on the 6.7 mayor release because this code only runs in tests for months. The aproach should be made more integrational.

aragon999 replied:
Mh, but how should the change of the type `StorefrontSearchResult` -> `EntitySearchResult` be implemented otherwise?

ssltg replied:
The `StorefrontSearchResult` already extends from `EntitySearchResult`, so there is no need for a hard rewrite on this step. The use of `StorefrontSearchResult` methods in 6.7 is already captured by the triggerDeprecation methods. So I would just leave the `ReviewLoaderResult` techically as it is and just add a deprecation notice, that it will be extended directly from `EntitySearchResult` from 6.7 on,

```suggestion
* deprecated tag:v6.7.0 - will be extended directly from `EntitySearchResult`.
class ReviewLoaderResult extends StorefrontSearchResult /* extends EntitySearchResult */
{
```

aragon999 replied:
@ssltg thank you for the feedback. But in general I meant more the part of the return types:
```php
    public function getReviews(): EntitySearchResult
    {
        return $this->reviews;
    }
```
vs
```php
    public function getReviews(): StorefrontSearchResult
    {
        return $this->reviews;
    }
```

Should I also do it with comments there? But than I do not know how to implement it in the rest of the code?

ssltg replied:
@aragon999 ok, I see your problem here. The reviews are a bit odd implemented here in the first place. Leave the code like it is. Sorry for the inconveniences.

aragon999 replied:
@ssltg okay very good :-) Yes it is also implemented the same way in other places, therefore I did not find another solution which was non-breaking.

Regarding the reviews, yes, I once tried to clean that a bit up, but it was also quite challenging: https://github.com/shopware/shopware/pull/2389/files

Maybe once this is merged, I can give this another try in the Shopware 6.6 cycle.

dfrancos-hub commented on phpstan-v67-baseline.neon:
```diff
@@ -0,0 +1,96 @@
+parameters:
+	ignoreErrors:
+		-
+			message: "#^Method \"getOrders\" of class \"Shopware\\\\Storefront\\\\Page\\\\Account\\\\Order\\\\AccountOrderPage\" is marked as deprecated, but does not call \"Feature\\:\\:triggerDeprecationOrThrow\"\\. All deprecated methods need to trigger a deprecation warning\\.$#"
```
Hi @aragon999, am pasting here comments from the internal core team review:

We should adding those triggerDeprecationOrThrow methods.

aragon999 replied:
@dfrancos-hub @ssltg does it really make sense to throw deprecation errors here? :thinking:

ssltg replied:
Yes, because we want to keep the phpstan exceptions as low as possible. I know in the current state it is technically not possible that the deprecation is triggered in these methods, but to trigger the deprecation errors helps us to make sure this will stay until 6.7. As many developers are working on our codebase, this is the best way to ensure compatibility even through automatic merge conflict resolves.

dfrancos-hub commented:
hi @aragon999 , the changes are merged 👍 . Thanks once again for the contribution 🙏
